### ISSUE TITLE
v3 new feature: undelete (database) object 

### ISSUE THREAD

**Issue by sun-geo** [openstreetmap/iD#6539]
A general core mapping idea is to hold and save the object history.
For example if you want to make existing building outlines more precise (maybe because there are background pictures with higher resolution available) it is recommend to use the existing outline and modify them instead of breaking the object-history by deleting & adding objects.
Similar behavior could happen if there are deleted objects by mistake (maybe accidently done from new osm-users) it would be nice to have an option to undelete these objects.

JOSM has a function which is called "undelete object" and is doing a nice job.
Please could you implement similar function?

**Comment by sun-geo** [comment]
Thank you for not closing this request/idea immediately and classifying it as "considering". :-)

Of course, I see this function not on the iD-main-window, so that the standard iD-user will not be bothered with too many options/functions. Maybe it could be located somewhere in expect/advanced section, or somewhere close to the raw-editor or even a that area where the raw-raw-editor (bulk tag copy/paste function) is doing operation.

By the way, would the implementation be quite crafty and time consuming?

**Comment by quincylvania** [comment]
@sun-geo I think the idea is great! I'm just not sure how or when this could fit into iD.

**Comment by mmd-osm** [comment]
Undeleting objects in JOSM isn't exactly straightforward: https://wiki.openstreetmap.org/wiki/JOSM/Plugins/Undelete says you need to supply an object id first, and identifying such an object id not that easy: https://wiki.openstreetmap.org/wiki/Find_the_id_of_a_deleted_node

Undelete support in the API is on the Top Ten Tasks list, btw (see https://wiki.openstreetmap.org/wiki/Top_Ten_Tasks#OSM_API_deleted_items_call). Without proper backend API support, this will be quite difficult to implement in a user friendly way.

**Comment by sun-geo** [comment]
> Undeleting objects in JOSM isn't exactly straightforward:

I agree, it isn't straightforward.

Maybe this is the first idea of a panel how the GUI for such a function could look like:

![undelete object](https://user-images.githubusercontent.com/47290585/70745485-2715f500-1d24-11ea-8170-46487cfd07d6.png)

- left side a field for inserting object ID, of the object you want to restore
- the usual "i"-button to send the iD user to your mentioned osm-wiki article (https://wiki.openstreetmap.org/wiki/Find_the_id_of_a_deleted_node)
- a drop down menu for Node / Way / Relation element selection
- a download button, to restore the object
- and a zoom-to-object button, for showing and zooming to the object
(the last 2 could also automatically combined)

The whole panel of course should be hidden by default, and somewhere could be a show button (or a show/hide switch)

**Comment by matkoniecz** [comment]
> Undeleting objects in JOSM isn't exactly straightforward: https://wiki.openstreetmap.org/wiki/JOSM/Plugins/Undelete says you need to supply an object id first, and identifying such an object id not that easy: https://wiki.openstreetmap.org/wiki/Find_the_id_of_a_deleted_node

Given how complex and confusing is to find id to undelete, I expect that such users would have no problem with using JOSM or other tools. 

Maybe it would make more sense to focus on making basic editing easier and less confusing in iD?

**Comment by tyrasd** [comment]
I'd agree with @matkoniecz and @mmd-osm on this one: This is a potentially really complex operation, and if a user is savvy enough to know what they are doing, it can be already done via JOSM and other tools.

**Comment by 1ec5** [comment]
I suppose most of the complexity comes from how the operation is (not) being exposed by the OSM API. Potlatch 1 used to have this feature (for ways only), which ordinary mappers had little problem using, but it depended on an alternative API format that has since been discontinued. So if anyone would still like to see a revival of this feature, it needs to be tracked in the openstreetmap-website repository before we can consider a UI for it here.

**Comment by matkoniecz** [comment]
I would say that most of complexity comes from it being a revert.

For example, what should happen if we undelete way that contained nodes that were deleted/moved since then?
What about nodes not edited?

And undeleting relations may be a nightmare of unlimited complexity.

**Comment by matkoniecz** [comment]
But better API may help.

**Comment by mmd-osm** [comment]
https://github.com/openstreetmap/openstreetmap-website/pull/1448 is the old discussion on this topic. It’s fairly unlikely this will happen anytime soon due to the semantic and heavy performance issues to be addressed.
